cpp/src/arrow/util/decimal_test.cc

```diff
@@ -1230,65 +1275,72 @@ TYPED_TEST(TestDecimalToRealDouble, LargeValues) {
+                                          -9.999999999999998e+47);
+  }
   // Integers are always exact
   auto scale = TypeParam::kMaxScale - 1;
   std::string seven = "7.";
   seven.append(scale, '0');
   CheckDecimalToReal<TypeParam, double>(seven, scale, 7.0);
   CheckDecimalToReal<TypeParam, double>("-" + seven, scale, -7.0);
-
-  CheckDecimalToReal<TypeParam, double>("99999999999999999999.0000000000000000", 16,
-                                        99999999999999999999.0);
-  CheckDecimalToReal<TypeParam, double>("-99999999999999999999.0000000000000000", 16,
-                                        -99999999999999999999.0);
+  if constexpr (TypeParam::kMaxPrecision >= 20) {
+    CheckDecimalToReal<TypeParam, double>("99999999999999999999.0000000000000000", 16,
+                                          99999999999999999999.0);
+    CheckDecimalToReal<TypeParam, double>("-99999999999999999999.0000000000000000", 16,
+                                          -99999999999999999999.0);
+  }
 
   // Small fractions are within one ULP
   CheckDecimalToRealWithinOneULP<TypeParam, double>("9999999.9", 1, 9999999.9);
   CheckDecimalToRealWithinOneULP<TypeParam, double>("-9999999.9", 1, -9999999.9);
-  CheckDecimalToRealWithinOneULP<TypeParam, double>("9999999.999999999999999", 15,
-                                                    9999999.999999999999999);
-  CheckDecimalToRealWithinOneULP<TypeParam, double>("-9999999.999999999999999", 15,
-                                                    -9999999.999999999999999);
-
-  // Large fractions are within 2^-52
-  constexpr double epsilon = 2.220446049250313080847263336181640625e-16;  // 2^-52
-  CheckDecimalToRealWithinEpsilon<TypeParam, double>(
-      "112334829348925.99070703983306884765625", 23, epsilon,
-      112334829348925.99070703983306884765625);
-  CheckDecimalToRealWithinEpsilon<TypeParam, double>(
-      "1.987748987892758765582589910934859345", 36, epsilon,
-      1.987748987892758765582589910934859345);
+  if constexpr (TypeParam::kMaxPrecision >= 23) {
+    CheckDecimalToRealWithinOneULP<TypeParam, double>("9999999.999999999999999", 15,
+                                                      9999999.999999999999999);
+    CheckDecimalToRealWithinOneULP<TypeParam, double>("-9999999.999999999999999", 15,
+                                                      -9999999.999999999999999);
+    // Large fractions are within 2^-52
+    constexpr double epsilon = 2.220446049250313080847263336181640625e-16;  // 2^-52
+    CheckDecimalToRealWithinEpsilon<TypeParam, double>(
+        "112334829348925.99070703983306884765625", 23, epsilon,
+        112334829348925.99070703983306884765625);
+    CheckDecimalToRealWithinEpsilon<TypeParam, double>(
+        "1.987748987892758765582589910934859345", 36, epsilon,
+        1.987748987892758765582589910934859345);
+  }
 }
 
 #endif  // __MINGW32__
 
-TEST(Decimal128Test, TestFromBigEndian) {
+TEST(Decimal32Test, TestFromBigEndian) {
```

Comment:
It seems a number of tests are mostly or exactly identical between the Decimal tests, perhaps you can write a generic version and call it for each concrete test type?

Reply:
added generic versions for most of them. the LeftShift/RightShift and Negate tests are sufficiently different in their values that it's harder to make them generic